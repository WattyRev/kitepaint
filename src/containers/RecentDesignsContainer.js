import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Design from "../models/Design";
import productShape from "../models/Product";
import Manufacturer from "../models/Manufacturer";
import { getRecentDesigns } from "../redux/modules/designs";
import { getProductsWithIndex } from "../redux/modules/products";
import { getManufacturersWithIndex } from "../redux/modules/manufacturers";
import { GET_DESIGNS, GET_PRODUCTS, GET_MANUFACTURERS } from "../redux/actions";
import { makeCancelable } from "../utils";

/**
 * Provides information and actions about/for Designs.
 */
export class RecentDesignsContainer extends React.Component {
  static propTypes = {
    /**
     * A function returning content to be rendered
     */
    children: PropTypes.func.isRequired,
    /**
     * A list of designs
     */
    designs: PropTypes.arrayOf(PropTypes.instanceOf(Design)).isRequired,
    /** The manufacturers indexed by id */
    manufacturers: PropTypes.objectOf(PropTypes.instanceOf(Manufacturer))
      .isRequired,
    /** The products indexed by id */
    products: PropTypes.objectOf(productShape).isRequired,
    /**
     * A function to trigger the retrieval of the designs. This should update the redux state,
     * causing designs to be provided through redux.
     */
    getDesigns: PropTypes.func.isRequired,
    /** A function to trigger the retrieval of all products */
    getProducts: PropTypes.func.isRequired,
    /** A funciton to trigger the retrieval of all manufacturers */
    getManufacturers: PropTypes.func.isRequired
  };

  state = {
    isLoading: true
  };

  componentDidMount() {
    // Get the designs, but limit the request to 6 since we don't need to show many.
    const designRequest = makeCancelable(
      this.props.getDesigns({
        limit: 6
      })
    );
    const productsRequest = makeCancelable(this.props.getProducts());
    const manufacturersRequest = makeCancelable(this.props.getManufacturers());
    this.cancelablePromises.push(designRequest);
    this.cancelablePromises.push(productsRequest);
    this.cancelablePromises.push(manufacturersRequest);

    Promise.all([
      designRequest.promise,
      productsRequest.promise,
      manufacturersRequest.promise
    ])
      .then(() => this.setState({ isLoading: false }))
      .catch(response => {
        if (response && response.isCanceled) {
          return;
        }
        this.setState({ isLoading: false });
      });
  }

  componentWillUnmount() {
    this.cancelablePromises.forEach(cancelable => cancelable.cancel());
  }

  cancelablePromises = [];

  render() {
    return this.props.children({
      props: {
        isLoading: this.state.isLoading,
        designs: this.props.designs,
        products: this.props.products,
        manufacturers: this.props.manufacturers
      }
    });
  }
}

const mapStateToProps = state => ({
  designs: getRecentDesigns(state, 6),
  products: getProductsWithIndex(state),
  manufacturers: getManufacturersWithIndex(state)
});

const mapDispatchToProps = {
  getDesigns: GET_DESIGNS,
  getProducts: GET_PRODUCTS,
  getManufacturers: GET_MANUFACTURERS
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RecentDesignsContainer);
