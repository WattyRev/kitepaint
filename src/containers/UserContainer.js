import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { getUser, getUserRecognition } from "../redux/modules/user";
import { SET_RECOGNIZED_USER, LOG_IN, LOG_OUT } from "../redux/actions";

/**
 * Maintains a local storage variable that indicates if the user is recognized as a registered user.
 * TODO this needs to be done in redux, because there will need to be more than one of these, and
 * they need to stay in sync.
 *
 * @param {Function} children
 */
class UserContainer extends React.Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    onSetRecognition: PropTypes.func.isRequired,
    onLogIn: PropTypes.func.isRequired
  };

  /**
   * Toggles the variable in state and in local storage
   */
  toggleRecognition = () => {
    const { isRecognizedUser } = this.props;
    this.props.onSetRecognition(!isRecognizedUser);
  };

  render() {
    const {
      firstName,
      id,
      isLoggedIn,
      isLoggingIn,
      lastName,
      username
    } = this.props.user;
    return this.props.children({
      actions: {
        toggleRecognition: this.toggleRecognition,
        logIn: this.props.onLogIn,
        logOut: this.props.onLogOut
      },
      props: {
        firstName,
        id,
        isLoggedIn,
        isLoggingIn,
        isRecognizedUser: this.props.isRecognizedUser,
        lastName,
        username
      }
    });
  }
}

const mapStateToProps = state => ({
  user: getUser(state),
  isRecognizedUser: getUserRecognition(state)
});

const mapDispatchToProps = {
  onLogIn: LOG_IN,
  onLogOut: LOG_OUT,
  onSetRecognition: SET_RECOGNIZED_USER
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UserContainer);
