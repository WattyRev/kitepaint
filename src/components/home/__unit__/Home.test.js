import React from "react";
import { shallow } from "enzyme";
import UserContainer from "../../../containers/UserContainer";
import RecentDesignsContainer from "../../../containers/RecentDesignsContainer";
import Home from "../Home";

describe("Home", () => {
  let defaultProps;
  beforeEach(() => {
    defaultProps = {};
  });
  it("renders", () => {
    expect.assertions(3);

    // Test the top level
    const wrapper = shallow(<Home {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();

    // Test the content inside UserContainer
    const userContainerData = {
      props: {
        isLoggedIn: false,
        isRecognizedUser: true
      },
      actions: {
        toggleRecognition: jest.fn()
      }
    };
    const userContainerWrapper = shallow(
      <div>
        {wrapper.find(UserContainer).prop("children")(userContainerData)}
      </div>
    );
    expect(userContainerWrapper).toMatchSnapshot();

    // Test the content inside RecentDesignsContainer
    const recentDesignsData = {
      props: {
        designs: [],
        manufacturers: {},
        products: {}
      }
    };
    const recentDesignsWrapper = shallow(
      <div>
        {wrapper.find(RecentDesignsContainer).prop("children")(
          recentDesignsData
        )}
      </div>
    );
    expect(recentDesignsWrapper).toMatchSnapshot();
  });
});
