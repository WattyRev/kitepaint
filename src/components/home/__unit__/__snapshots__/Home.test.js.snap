// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders 1`] = `
<Fragment>
  <CTABanner />
  <Connect(UserContainer)>
    <Component />
  </Connect(UserContainer)>
  <Connect(RecentDesignsContainer)>
    <Component />
  </Connect(RecentDesignsContainer)>
</Fragment>
`;

exports[`Home renders 2`] = `
<div>
  <AccountBanner
    isRecognizedUser={true}
  >
    <AccountForm
      id="account-banner-login"
      isRecognizedUser={true}
      onToggleRecognition={[MockFunction]}
    />
  </AccountBanner>
</div>
`;

exports[`Home renders 3`] = `
<div>
  <RecentDesignsBanner
    designs={Array []}
  />
</div>
`;
